src/Internal/Types/ScriptLookups.purs

```diff
@@ -1028,13 +1034,20 @@ processConstraint mpsMap osMap = do
           , data: unwrap red
           , exUnits: zero
           }
-      -- Remove mint redeemers from array before reindexing.
-      _redeemersTxIns %= filter \(T.Redeemer { tag } /\ _) -> tag /= Mint
-      -- Reindex mint redeemers.
-      mintRedeemers <- lift $ reindexMintRedeemers mpsHash redeemer
-      -- Append reindexed mint redeemers to array.
-      _redeemersTxIns <>= map (_ /\ Nothing) mintRedeemers
-      _cpsToTxBody <<< _mint <>= map wrap mintVal
+
+      let mp = lookupMintingPolicy mpsHash mpsMap
+      case mp of
```

Comment:
I think we should use `mp <- except $ lookupMintingPolicy`, because the `Left` case is an exception. It will be handled a few lines above, but still I'd prefer the code style without wildcard patterns.

Reply:
It was `except $ ...` originally, but there is a case in the `ReferenceInputs` example where the hash of the minting policy is not in `mpsMap` and so `except` will throw an error. Although, I agree that it can be refactored somehow to use the pattern match above.

Reply:
But it is possible that the minting policy is not in `mpsMap` simply because the user forgot to add it to lookups, right? And in this case we will not be able to detect that the intention was to use a `NativeScript`, so the failure will be the same as before this change. Not a great UX.

If so, all of it becomes a bit hairy. I think we should introduce another set of constraints for native scripts then, since the logic is different in non-trivial way.

If we have to provide the native script anyways for minting, let's pass it directly, without providing a hash.

So,

```
mustMintCurrencyUsingNativeScript
  :: forall (i :: Type) (o :: Type)
   . NativeScript
  -> TokenName
  -> BigInt
  -> TxConstraints i o
```

Please also add a new `TxConstraint` constructor for minting using native scripts.

My motivation:

I think we should introduce another constraint sooner, because in the future we'd like to remove lookups, and modifying `mustMintCurrency` so that it is able to handle any script will require something like:

```
mustMintCurrency
  :: forall (i :: Type) (o :: Type)
   . Either PlutusScript NativeScript
  -> TokenName
  -> BigInt
  -> TxConstraints i o
```

which is really ugly (because of the need to tag with `Either`).

Reply:
One more argument for separating the mint constraint into native/plutus variants: we have already done the same for `mustPayTo*Script`

Reply:
Yes this separation makes sense,  I'll do the refactor.